### PR TITLE
DMP-2402: Upload audio to azure directly from DARTS gateway

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
Transferring audio files between the Proxy and the API for upload to Azure was observed to be slow during performance testing. To address, the proxy should upload the audio file to Azure blob store directly, instead of sending over to the API. In case of failure, the metadata will be stored in Azure blob store, associated to the audio file. Then, the new /audio/metadata endpoint (DMP-4115) in DARTS API should be invoked from the Proxy to save the metadata of the audio file to the database. 

This change should also be applied to the Gateway. If this is not trivial to just copy the code across please raise a new ticket to track. 
h3. Acceptance criteria

*AC1: Send audio and metadata into DARTS Proxy*

GIVEN an audio file is sent

WHEN it is received into the Proxy

THEN the audio file and metadata is uploaded to Azure Blob store

AND the metadata is sent to the DARTS API and associated with the correct case

 

*AC2: Send audio and metadata into DARTS Gateway*

GIVEN an audio file is sent

WHEN it is received into the Gateway

THEN the audio file and metadata is uploaded to Azure Blob store

AND the metadata is sent to the DARTS API and associated with the correct case

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -44,6 +44,18 @@ spring:
...
 azure:
 darts-gateway:
+  storage:
+    blob:
+      client:
+        connection-string: ${AZURE_STORAGE_CONNECTION_STRING:}
+        ## See https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blobs-tune-upload-download-java
+        block-size-bytes: 10_000_000
+        max-single-upload-size-bytes: 10_000_000
+        max-concurrency: 5
```
Comment (cmagowan-hmcts) [inline src/main/resources/application.yaml:54]: is this restricting the number of uploads to blob storage at once?
if we have 155 courthouses, would this need to be bigger?
@hemantasharma1129

**Review by cmagowan-hmcts — APPROVED** [review]

File: bin/secrets-stg-environment.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+# To set the secrets in your shell, source this file ie. source ./bin/secrets-stg.sh
```
Comment (davet1985) [inline bin/secrets-stg-environment.sh:3]: The file name is wrong here

File: build.gradle [hunk 3]
```diff
@@ -224,457 +224,458 @@ project.tasks['check'].finalizedBy integration
...
-  'src/main/java/generated/**',
-  '**/uk/gov/hmcts/darts/**/model/**',
-  '**/uk/gov/hmcts/darts/**/config/**',
-  '**/enums/**',
-  '**/DocumentumIdToJwtCache*',
-  '**/CacheValueWithJwt*',
-  '**/com/service/viq/event/**',
-  '**/uk/gov/hmcts/darts/cache/token/service/**',
-  '**/uk/gov/hmcts/darts/test/**'
+        '**/com/emc/**',
```
Comment (davet1985) [inline build.gradle:227]: Why the change in indentation across the whole file? I think this change should be removed.

File: src/integrationTest/resources/application-int-test-jwt-token-shared.yaml [hunk 4]
```diff
@@ -10,6 +10,10 @@ logging:
     root: ERROR
 
 darts-gateway:
+#  storage:
+#    blob:
+#      client:
+#        disable-upload: true
```
Comment (davet1985) [inline src/integrationTest/resources/application-int-test-jwt-token-shared.yaml:16]: Remove comments?

File: src/main/java/uk/gov/hmcts/darts/common/client/AudiosClient.java [hunk 5]
```diff
@@ -56,4 +62,20 @@ public ResponseEntity<Void> addAudio(MultipartFile file, AddAudioMetadataRequest
...
         return new ResponseEntity<>(HttpStatusCode.valueOf(200));
     }
+
+    @Override
+    @SuppressWarnings("PMD.LooseCoupling")
+    public ResponseEntity<Void> addAudioMetaData(AddAudioMetadataRequestWithStorageGUID addAudioMetadataRequestWithStorageGuid) {
+        try {
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentType(MediaType.APPLICATION_JSON);
+            headers.setAccept(List.of());
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/common/client/AudiosClient.java:72]: Do we need to do this?

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -44,6 +44,18 @@ spring:
...
+        timeout: 15m
+        disable-upload: false
+      container-name:
+        inbound: darts-inbound-container
```
Comment (davet1985) [inline src/main/resources/application.yaml:58]: Do we think these configs might need to be overridable in flux? CC @jackmaloney @hemantasharma1129

File: src/test/java/uk/gov/hmcts/darts/utilities/DataUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        addAudioMetadataRequest.setCourtroom("courtroom");
+        addAudioMetadataRequest.setMediaFile("mediaFile");
+        addAudioMetadataRequest.setFileSize(100L);
+        addAudioMetadataRequest.setChecksum("checksum");
+        addAudioMetadataRequest.setCases(List.of("case1", "case2"));
+        return addAudioMetadataRequest;
+    }
+
+    @Test
+    void positiveToMap() {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/utilities/DataUtilTest.java:35]: Do we need any negative tests here?

**Review by davet1985 — CHANGES_REQUESTED** [review]
I've added a few questions/comments.

File: src/main/resources/application.yaml (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:54]: @hemantasharma1129  

max-single-upload-size-bytes is just the maximum size per upload if a file is larger then this value it will be split up into chunks and uploaded in chunks. But if it is less then this then it will be uploaded in one go.
image

I believe max-concurrency relates to the maximum threads the chunked upload can use and is applied per upload not global.
image
 





These settings were pulled from darts-api :)

File: bin/secrets-stg-environment.sh (hunk 2 above)
Comment (Ben-Edwards-cgi) [reply bin/secrets-stg-environment.sh:3]: Updated

File: build.gradle (hunk 3 above)
Comment (Ben-Edwards-cgi) [reply build.gradle:227]: The indentation was auto added by the HMCTs code formatter config. 
As such I believe this better aligns to the HMCTs standards :)

File: src/integrationTest/resources/application-int-test-jwt-token-shared.yaml (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/integrationTest/resources/application-int-test-jwt-token-shared.yaml:16]: Removed :)
This is left over test code

File: src/main/java/uk/gov/hmcts/darts/common/client/AudiosClient.java (hunk 5 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/client/AudiosClient.java:72]: Yes :)

If we don't reset the accept flag, it will be given application/json by default which will be rejected by the darts-api 
 as this API does not have a response body as such it fails to find an API that can return application/json

File: build.gradle (hunk 3 above)
Comment (davet1985) [reply build.gradle:227]: Ok that's fine. It's 985 changed lines that are not related to the other changes in this PR, which makes it unnecessarily noisy for reviewing.

File: src/test/java/uk/gov/hmcts/darts/utilities/DataUtilTest.java (hunk 7 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/utilities/DataUtilTest.java:35]: I have added some null checks to ensure we don't break if the object is missing data (but this "should" be filtered out when we receive the API but its always good from a belts and braces viewpoint to verify)

File: src/main/resources/application.yaml (hunk 6 above)
Comment (davet1985) [reply src/main/resources/application.yaml:58]: Ok, I have created a ticket to review these across both services - [DMP-4317](https://tools.hmcts.net/jira/browse/DMP-4317)

**Review by davet1985 — APPROVED** [review]
